common/gamelib/movement-manager.mjs

```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: BSD-2-Clause
+
+import {checkPlayerMove, getPlayerMoveCooldown} from './move-check.mjs';
+import {PlayerSyncMessage} from './player.mjs';
+import {MapCoord} from '../maplib/map.mjs';
+
+const ERROR_COOLDOWN_MS = 300;
+
+/**
+ *TODO: jsdoc
+ */
+class MovementManagerServer {
+  /**
+   * TODO: jsdoc
+   */
+  constructor() {
+    this.playerMovingTimer = new Map();
+  }
+
+  /**
+   * TODO: jsdoc
+   * @param {GateWayService} gs
+   * @param {Socket} socket
+   * @param {PlayerSyncMessage} updateMsg
+   * @param {Function} failOnPlayerUpdate
+   */
+  async handlePlayerMove(gs, socket, updateMsg, failOnPlayerUpdate) {
+    const player = socket.playerData;
+
+    // If the update message comes too early, handles it later.
+    if (getPlayerMoveCooldown(player) > 0) {
+      // We reserve only one moving request per player.
+      if (this.playerMovingTimer.has(player.playerID)) {
+        clearTimeout(this.playerMovingTimer.get(player.playerID));
+      }
+      // Handle the moving request later.
+      setTimeout((gs, socket, updateMsg, failOnPlayerUpdate) => {
+        this.handlePlayerMove(gs, socket, updateMsg, failOnPlayerUpdate);
+      }, Math.max(getPlayerMoveCooldown(player), 0), gs, socket, updateMsg, failOnPlayerUpdate);
+      return;
+    }
+
+    // clean up the timer
+    this.playerMovingTimer.delete(player.playerID);
+
+    if (!checkPlayerMove(player, updateMsg, gs.gameMap)) {
+      failOnPlayerUpdate();
+      return;
+    }
+
+    if (!(await gs._teleportPlayerInternal(socket, updateMsg, false))) {
+      failOnPlayerUpdate();
+      return;
+    }
+  }
+}
+
+
+/**
+ * TODO: jsdoc
+ */
+class MovementManagerClient {
+  /**
+   * Construct a movement manager
+   * @constructor
+   * @param {Socket} socket - A socket.io socket.
+   * @param {GameState} gameState
+   * @param {InputManager} inputManager
+   * @param {MapRenderer} mapRenderer
+   */
+  constructor(socket, gameState, inputManager, mapRenderer) {
+    this.socket = socket;
+    this.gameState = gameState;
+    this.inputManager = inputManager;
+    this.mapRenderer = mapRenderer;
+
+    this.gameClient = null;
+    window.addEventListener('dataReady', (event) => {
+      this.gameClient = event.detail.gameClient;
+
+      this.inputManager.registerMapMove((direction) => {
+        const dx = {'U': 0, 'D': 0, 'L': -1, 'R': 1}[direction];
+        const dy = {'U': 1, 'D': -1, 'L': 0, 'R': 0}[direction];
+        const {x, y} = this.gameClient.playerInfo.mapCoord;
+        this.moveTo(x + dx, y + dy, direction, inputManager.pressedKeys.has('g'));
+      });
+
+      // ghost mode
+      this.inputManager.registerKeydownOnce(mapRenderer.canvas, (event) => {
+        if (event.key !== 'g') return;
+        const {mapCoord: {x, y}, facing} = this.gameClient.playerInfo;
+        this.moveTo(x, y, facing, true);
+      });
+      this.inputManager.registerKeyup(mapRenderer.canvas, (event) => {
+        if (event.key !== 'g') return;
```

Comment:
Optional: Keyup may not fire if user alt-tab away half way.

Reply:
It is input manager's job to send a keyup event in that situation.

(although that is still a TODO ...
https://github.com/canyugs/hitcon-online/blob/74624461de1badf58460923d44cbeadbaac72f8a/sites/game-client/input-manager.mjs#L56